### ISSUE TITLE
"Connection security" notification when on login-wall wifi

### ISSUE THREAD

**Issue by jancborchardt** [stefan-niedermann/nextcloud-notes#426]
When in an airport or on the train (or any other sort of wifi with a login-wall) and are connected but not logged in to that wifi, you get this very technical notification:
![](https://i.imgur.com/R5ITL7C.jpg)

Instead, Notes should simply continue to use the mobile network and not show a notification.

**Comment by schaarsc** [comment]
please don't use mobile network if WIFI is connected.

**Comment by jancborchardt** [comment]
@schaarsc what do you mean by that? Is that how the app could fix that? I wouldn’t want to require from everyday users to need to disable their mobile network whenever wifi is connected.

Besides, the problem is specifically that wifi _is not_ connected, but in the phase where you are connected but not logged in through the log in wall. So actually, disabling the wifi and using your mobile network resolves the issue, but eats your data plan.

@stefan-niedermann any idea? :) Maybe @AndyScherzinger @tobiasKaminsky from the Nextcloud Android app have input how to handle this without technical notifications? :)

**Comment by tobiasKaminsky** [comment]
Normally Android should be smart enough to check this and then switch to mobile data.
Maybe the connection try was too short after establishing the wifi connection.
Do the next connections try also not work?

**Comment by korelstar** [comment]
That notification is created by the library cert4android which gives us support for self-signed certificates. The behavior would change, if we use the Nextcloud SSO library for Android (see #418).

However, I don't think that an Android developer can dictate the connection type (I.e. WLAN or mobile data). AFAIK, this is done by the OS.

Is the files app able to connect to the server in that state or does it just not show an error for this situation?

**Comment by mario** [comment]
@korelstar you can select app-wide to use either wifi or mobile if you're connected to both.

**Comment by schaarsc** [comment]
if I have WIFI enabled I expect all my traffic to go through the WIFI connection. If the WIFI connection has a problem I want to be informed about the problem so that I can solve the issue. But I don't want any App to use mobile connections while WIFI is enabled.

**Comment by jancborchardt** [comment]
I’m not sure the issue is clear here:

- This happens when you are about to log in to a login-wall-wifi, like on a train or an airport, or maybe a café
- Android itself shows you that you should "Sign in to Wi-Fi Network" as seen in the screenshot above
- No other app shows anything, except Notes (and DAVdroid some time ago). And what it showed is a very technical error no regular person understands.

It is not our task to handle this. Our task is to keep people’s data in sync whatever the situation is.

**Comment by schaarsc** [comment]
@jancborchardt in your original post you requested a fall back to the mobile network. I'm against this fall back without user consent because users (at least me) expect to be using the WIFI as soon as the WIFI symbol appears in the status bar. 

Options to avoid certificate warning:
1. disable validation, not recommended
2. (requires API level 21) register https://developer.android.com/reference/android/net/ConnectivityManager.NetworkCallback and do not send requests if https://developer.android.com/reference/android/net/NetworkCapabilities#NET_CAPABILITY_CAPTIVE_PORTAL is detected
3. (requires API level 23) Allow users to log in from Notes App using https://developer.android.com/reference/android/net/ConnectivityManager#ACTION_CAPTIVE_PORTAL_SIGN_IN

Any other ideas?

I have not tested this, but I think option 2. should be feasible and not to much work. Would this solve the issue for you?

**Comment by schaarsc** [comment]
Option 4
Notes app min API level is currently 14. To keep that min level we could try to detect the captive portal in the app by connecting to some known host and check for redirects (see https://developer.android.com/reference/kotlin/java/net/HttpURLConnection).

**Comment by korelstar** [comment]
Don't re-implement the wheel (don't realize option 4)!

In order to avoid raising the minimum API level for the whole app, check the current level at runtime (pseudo-code):

```
if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP && !NET_CAPABILITY_INTERNET) {
cancelsync;
}
```

**Comment by schaarsc** [comment]
@korelstar are you suggesting to fix this issue only for API>=21?
when are we planing to raise the minApiLevel anyway? 

NB: option 4 works as expected. but since we don't have a callback in api<21 the app has to be closed to trigger updateNetworkStatus()

**Comment by schaarsc** [comment]
I've implemented option 2. and option 4.
It appears that option 4 is more reliable. At least my device did not detect the captive portal (tested with two different hotspots)
please let me know if you are interested in a PR.

**Comment by korelstar** [comment]
> @korelstar are you suggesting to fix this issue only for API>=21?

Yep, I think that's fine, since it's "only" a UX issue. Raising the API level to 21 would exclude those people completely.

I don't like option 4, since it's potentially an issue of data privacy/protection or reliability (which host do you connect for those tests?).

**Comment by schaarsc** [comment]
I'm using DEFAUL_SERVER from http://androidxref.com/6.0.0_r1/xref/frameworks/base/packages/CaptivePortalLogin/src/com/android/captiveportallogin/CaptivePortalLoginActivity.java#59

**Comment by jancborchardt** [comment]
@schaarsc the premise is not true though:

> because users (at least me) expect to be using the WIFI as soon as the WIFI symbol appears in the status bar.

The wifi symbol appears, but with a big **"!"** next to it. Additionally there is a notification saying that you need to sign in. People don’t need any more info from a specific app.

For the implementation, 2 things are important:
- Data should be synced at all times. If you are not connected to a wifi yet (as in this case) mobile data should be used, just as it was before "connecting" to the wifi.
- No additional notification or separate thing to confirm for people.

If an implementation you list fulfills that it sould be fine.

**Comment by schaarsc** [comment]
In the situation described above If I use a browser I cannot connect to the Internet. I'm blocked by the captive portal and the browser is not falling back to mobile. I don't see why an application should behave differently.

**Comment by jancborchardt** [comment]
@schaarsc ok, maybe this depends on the device or exact system. In any case, there should be no notification – this is really the biggest issue, as mentioned in the original post.

**Comment by schaarsc** [comment]
#418 will also solve this. I suggest to close this issue.

**Comment by jancborchardt** [comment]
#418 (Single sign on) seems to have nothing to do with captive portals for public wifis.

@stefan-niedermann @tobiasKaminsky @AndyScherzinger @korelstar @mario how do we best proceed in fixing this? :)

**Comment by tobiasKaminsky** [comment]
No idea to be honest. Android should automatically avoid using those restricted wifis, unless the user got access to it:
- connect to wifi on ice (deutsche bahn)
- see notification about capative login
- ignore this
- I am still able to browse NC or websites as I have mobile data
- only after accepting ToS of wifi, the wifi is used.

So this is, at least in some cases, handled by Android correctly.

Do you know of many people having this? Maybe it is only some strange implementation of a vendor?

**Comment by jancborchardt** [comment]
I basically have this every time I go by train or bus or am in a hotel or café or airport with these captive portals.

Same for GNOME Online Accounts and our desktop client, which is similarly annoying and confusing: https://github.com/nextcloud/desktop/issues/81

**Comment by stefan-niedermann** [comment]
Closing in favor of #418 .
